pan-domain-auth-play_2-6/src/main/scala/com/gu/pandomainauth/action/Actions.scala

```diff
@@ -0,0 +1,336 @@
+          .withSession(session = request.session - ANTI_FORGERY_KEY - LOGIN_ORIGIN_KEY)
+      } else {
+        showUnauthedMessage(invalidUserMessage(claimedAuth))
+      }
+    }
+  }
+
+  def processLogout(implicit request: RequestHeader) = {
+    flushCookie(showUnauthedMessage("logged out"))
+  }
+
+  def readAuthenticatedUser(request: RequestHeader): Option[AuthenticatedUser] = readCookie(request) map { cookie =>
+    CookieUtils.parseCookieData(cookie.value, settings.publicKey)
+  }
+
+  def readCookie(request: RequestHeader): Option[Cookie] = request.cookies.get(PublicSettings.assymCookieName)
+
+  def generateCookies(authedUser: AuthenticatedUser): List[Cookie] = List(
+    Cookie(
+      name = PublicSettings.cookieName,
+      value = LegacyCookie.generateCookieData(authedUser, settings.secret),
+      domain = Some(domain),
+      secure = true,
+      httpOnly = true
+    ),
+    Cookie(
+      name = PublicSettings.assymCookieName,
+      value = CookieUtils.generateCookieData(authedUser, settings.privateKey),
+      domain = Some(domain),
+      secure = true,
+      httpOnly = true
+    )
+  )
+
+  def includeSystemInCookie(authedUser: AuthenticatedUser)(result: Result): Result = {
+    val updatedAuth    = authedUser.copy(authenticatedIn = authedUser.authenticatedIn + system)
+    val updatedCookies = generateCookies(updatedAuth)
+    result.withCookies(updatedCookies: _*)
+  }
+
+  def flushCookie(result: Result): Result = {
+    val clearCookie = DiscardingCookie(
+      name = PublicSettings.cookieName,
+      domain = Some(domain),
+      secure = true
+    )
+    val clearAssymCookie = DiscardingCookie(
+      name = PublicSettings.assymCookieName,
+      domain = Some(domain),
+      secure = true
+    )
+    result.discardingCookies(clearCookie, clearAssymCookie)
+  }
+
+  /**
+    * Extract the authentication status from the request.
+    */
+  def extractAuth(request: RequestHeader): AuthenticationStatus = {
+    readCookie(request).map { cookie =>
+      PanDomain.authStatus(cookie.value, settings.publicKey) match {
+        case Expired(authedUser) if authedUser.isInGracePeriod(apiGracePeriod) =>
+          GracePeriod(authedUser)
+        case authStatus @ Authenticated(authedUser) =>
+          if (cacheValidation && authedUser.authenticatedIn(system)) authStatus
+          else if (validateUser(authedUser)) authStatus
+          else NotAuthorized(authedUser)
+        case authStatus => authStatus
+      }
+    } getOrElse NotAuthenticated
+  }
+
+  /**
+    * Action that ensures the user is logged in and validated.
+    *
+    * This action is for page load type requests where it is possible to send the user for auth
+    * and for them to interact with the auth provider. For API / XHR type requests use the APIAuthAction
+    *
+    * if the user is not authed or the auth has expired they are sent for authentication
+    */
+  object AuthAction extends ActionBuilder[UserRequest, AnyContent] {
```

Comment:
Is the only that that has changed is the `ActionBuilder` syntax?

I'm wondering if there's a way of avoiding so much copy-paste in the code between 2.5 and 2.6, especially since it's security-related stuff.